### PR TITLE
Make `DragGestureRecognizer` abstract methods public

### LINKED ISSUE #151446: Make `DragGestureRecognizer` abstract methods public
### Use case

`DragGestureRecognizer` relies on abstract methods that `VerticalDragGestureRecognizer`, `HorizontalDragGestureRecognizer` and `PanGestureRecognizer` implement.

Such methods include `_considerFling`,  `_hasSufficientGlobalDistanceToAccept`, among others.

Since these methods are private, developers can't extend `DragGestureRecognizer` to create their own recognizers. For example, a developer might want to modify the distance required to accept the gesture. Currently, developers need to copy the whole `DragGestureRecognizer` to modify that.

### Proposal

Make the abstract methods public. Also, provide public getters to the private members that the subclasses access, like `_finalPosition` and `_globalDistanceMoved`.

### PR DESCRIPTION
**Pull request by angelosilvestre**
Resolves #151446

`DragGestureRecognizer` defines several private abstract methods that are implemented by its subclasses.
In the **super_editor** package, we'd like to extend `PanGestureRecognizer` to make it more aggressive, so it can win the gesture arena when placed inside a `CustomScrollview`. However, since we can't override private methods, tweaking this single function would involve copying the entire `DragGestureRecognizer` interface and its `PanGestureRecognizer` implementation.

<br>

Methods that were updated in this PR:

| Method | Rationale |
|---|---|
| `_hasSufficientGlobalDistanceToAccept` | This is the most important method for us. Overriding this method allows tweaking the PanGestureRecognizer to be more aggressive. |
| `_considerFling` | In **super_editor** we use the PanGestureRecognizer, but we want the fling gesture to behave as if it was a VerticalDragRecognizer. We'll use the fling gesture just to scroll vertically. |
| `_finalPosition` | I added a getter to be able to access it inside `_considerFling`. |
| `_globalDistanceMoved` | I added a getter to be able to access it inside `_hasSufficientGlobalDistanceToAccept`. |

### CODE REVIEW

**Comment by angelosilvestre** [comment]
I'm not sure which tests could be added here. Since this PR does not change any functionality maybe it doesn't need a test?

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/gestures/monodrag.dart:341]: Maybe we could improve the API signature as part of this process!

```dart
Axis? get primaryDragAxis => null;
```

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/gestures/monodrag.dart:324]: I feel about 80% confident that this is a fantastic improvement to the Flutter framework, but a small concern is that once something becomes public, the potential amount of future breaking changes increases.

What is the advantage of subclassing `DragGestureRecognizer` rather than extending `OneSequenceGestureRecognizer` directly?

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/gestures/monodrag.dart:1]: > I'm not sure which tests could be added here.

Yeah, that's valid. Feel free to request a test exemption as per the [bot comment](https://github.com/flutter/flutter/pull/151627/#issuecomment-2224231812), or if you'd like, you could add a custom `DragGestureRecognizer` subclass to [monodrag_test.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/test/gestures/monodrag_test.dart) and add some tests to double-check that it works as expected.

**Review by nate-thegrate — COMMENTED** [review]
Hi @angelosilvestre, thanks for making this contribution!

I gave a bit of feedback, so take a look when you have a chance 🙂

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:324]: It's true that `OneSequenceGestureRecognizer` can be extended directly. However, for small changes, such as the one I described in the PR description, where we just want to customize the `PanGestureRecognizer` a little bit. 

It looks overwhelming to me that we need to copy the whole `DragGestureRecognizer` just to tweak `PanGestureRecognizer` a little bit.

What do you think?

Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:325]: Yep, that sounds fair. Thanks!

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: Here's an idea: since the goal is to create an aggressive pan gesture that beats a `CustomScrollView` in the arena, why don't you add a test that checks this behavior?

Then, if at some point someone makes a change that breaks this functionality, it would cause a test failure rather than potentially going unnoticed.

**Review by nate-thegrate — COMMENTED** [review]
Thanks very much for the update, this PR is looking fantastic!

I'll be happy to approve once we get a testing plan sorted out 👍

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: Sounds good. I'll give it a shot.

Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: @nate-thegrate I added a test. I pumped a tree starting all the way up from the `MaterialApp`.  Please let me know if I need to change it.

Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: Awesome!

All we need now is 1 more PR approval :)

**Review by nate-thegrate — APPROVED** [review]
Flutter_LGTM

Thanks for putting in the work here!

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: Thanks. Do you happen to know who could be tagged for the second review?

Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:1]: Unfortunately, the timing with this PR isn't great—we usually have weekly triage meetings, but this week got canceled since people are on vacation.

I'll try sending a message out once the weekend's over, and hopefully we can get a second reviewer within the next few days.

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/gestures/monodrag.dart:15]: Sorry, one more thing I completely glossed over: we usually do relative imports for files within the same package, and package imports otherwise.

```suggestion
import 'package:flutter/painting.dart' show Axis;
import 'package:flutter/scheduler.dart';

```

Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart]: Updated.

**Comment by nate-thegrate** [comment]
It looks like there are a couple of test failures: one of them was caused by changes outside this branch, and the other is a dependency loop (that would have gone undetected if we'd kept the relative import).

Using the `Axis` enum in `DragGestureRecognizer` has caused **gestures.dart** and **painting.dart** to be codependent—in fact, I bet this is why `_DragDirection` was created in the first place!

In order to resolve this, we could revert to using a public `DragDirection` enum (or maybe just `bool horizontal`).

Or we could migrate some code to prevent the loop. Since [**painting/basic_types.dart**](https://github.com/flutter/flutter/blob/82b63ff27d7a15581837c84ca9304d3023934545/packages/flutter/lib/src/painting/basic_types.dart#L60) is already exporting a type from **foundation.dart**, I think it'd be super straightforward to just move the enums in that file into [**foundation/basic_types.dart**](https://github.com/flutter/flutter/blob/82b63ff27d7a15581837c84ca9304d3023934545/packages/flutter/lib/src/foundation/basic_types.dart).

<br>

@angelosilvestre Sorry that the act of making a few things public is turning out to be more complex than anticipated. I don't think there's anything you need to do at this point; I'll reach out to some folks tomorrow and hopefully we can come up with a solid action plan.

**Comment by angelosilvestre** [comment]
@nate-thegrate Thanks for the update. Please let me know if there is anything I should do.

**Comment by nate-thegrate** [comment]
I created #151771 in order to hopefully unblock this PR. I guess we'll see how it goes!

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (Piinks) [inline packages/flutter/lib/src/gestures/monodrag.dart:293]: Would lastPosition be a better name here?

**Review by Piinks — COMMENTED** [review]
> Such methods include _considerFling, _hasSufficientGlobalDistanceToAccept, among others.

Can you update the PR description to list all of the methods that are being made public, and why? I am not sure all of them need to be exposed in this way.

File: packages/flutter/test/gestures/monodrag_test.dart
Comment (Piinks) [inline packages/flutter/test/gestures/monodrag_test.dart:265]: What about all of the other methods?

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart]: Yeah, probably. Updated.

File: packages/flutter/test/gestures/monodrag_test.dart
Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:265]: To make the test simpler I didn't included the other methods. This test doesn't care about the fling gesture, so it wouldn't help here.

Comment (nate-thegrate) [reply packages/flutter/test/gestures/monodrag_test.dart:265]: I think the idea behind this comment was: each method that we make public should be included in the test framework.

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:265]: Understood. I'll update it.

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Updated. 

For our use-case, `primaryDragAxis` could be kept private, since it has a default implementation and we don't need to change it. However, if someone needs to write a custom `VerticalDragRecognizer` or a `HorizontalDragRecognizer` they will need to override `primaryDragAxis`.

Comment (Piinks) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: I think if only some of the methods are required for your use case, we should only expose those needed, rather than exposing all of them.

https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#lazy-programming

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Updated. Could you please take a look again @Piinks ?

Comment (nate-thegrate) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Hopefully it was all right for me to edit the original PR comment; I think that table makes it a bit easier to visually parse.

My guess is that a second approval will be contingent on re-privatizing `getDeltaForDetails` and `getPrimaryValueFromOffset`, since they aren't needed for your use case.

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: I think that if those methods should be re-privatized, a default implementation should be provided. Otherwise, it won't be possible to subclass `DragGestureRecognizer` outside of the flutter package.

Comment (nate-thegrate) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Good point! I can see that as a reason to keep them as public.

…or, we could pull a ***pro gamer move*** and turn `DragGestureRecognizer` into a `sealed class`. This could help streamline the logic within the framework, and since extending `DragGestureRecognizer` directly would no longer be possible, developers wouldn't waste any time trying to do it and could instead extend its subclasses (or the base `OneSequenceGestureRecognizer` class).

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Would it be possible to extend a class that extends a sealed class? I've never done that. If it works, it works for me that way, we could extend `PanGestureRecognizer`. I'd like to hear @Piinks opinion on that before making this change.

Comment (nate-thegrate) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: > Would it be possible to extend a class that extends a sealed class?

Yes sir! The main goal of a `sealed` class is to enable exhaustive pattern-matching:

```dart
void foo(DragGestureRecognizer recognizer) {
  switch (recognizer) {
    case HorizontalDragGestureRecognizer():
      // ...
    case VerticalDragGestureRecognizer():
      // ...
    case PanGestureRecognizer():
      // ...
  }
}
```

Fortunately, passing a subtype of `PanGestureRecognizer` doesn't break this behavior!

Comment (angelosilvestre) [reply packages/flutter/test/gestures/monodrag_test.dart:268]: Sounds good, I'll give it a try.

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (nate-thegrate) [inline packages/flutter/lib/src/gestures/monodrag.dart:291]: Speaking of documentation… maybe we could expand on this one a bit so it more closely aligns with our [style guidelines](https://github.com/flutter/flutter/blob/05f294c7116f30a93195b479d55539f457edd83f/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-useless-documentation).

Possible things to mention:
- It's stored as a local/global `OffsetPair`
- It's used to create `DragEndDetails`

Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:291]: How about

```dart
/// The local and global offsets of the last pointer event received.
///
/// It is used to create the [DragEndDetails], which provides information about
/// the end of a drag gesture.
OffsetPair get lastPosition => _lastPosition;
```

Is there anything you think it could be improved?

Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:291]: Nope, everything LGTM after this change!

**Review by Piinks — COMMENTED** [review]
Apologies for the delayed response, I was away for a bit. Thanks for the updates!

File: packages/flutter/lib/src/gestures/monodrag.dart
Comment (Piinks) [inline packages/flutter/lib/src/gestures/monodrag.dart:78]: Looks like we did something similar in BaseTapAndDragGestureRecognizer. @Renzo-Olivares what do you think?

Comment (Renzo-Olivares) [reply packages/flutter/lib/src/gestures/monodrag.dart:78]: I think this is safe to do. My initial though was that it might not be since `DragGestureRecognizer` has been around for a while, and sealing it now would break anyone who is extending it. Looking through github https://github.com/search?q=%22extends+DragGestureRecognizer+%7B%22+path%3A*.dart&type=code&p=1 this doesn't seem to be the case though. All the usages I could find at least where just forks of the framework. Internally I couldn't find anyone extending `DragGestureRecognizer` either.

Comment (angelosilvestre) [reply packages/flutter/lib/src/gestures/monodrag.dart:78]: > My initial though was that it might not be since DragGestureRecognizer has been around for a while, and sealing it now would break anyone who is extending it

I don't think anyone is able to extend it right now, since there are private abstract methods that must be implemented.

Comment (nate-thegrate) [reply packages/flutter/lib/src/gestures/monodrag.dart:78]: Now there's a real-world use case for https://github.com/dart-lang/linter/issues/4984!

**Review by Renzo-Olivares — APPROVED** [review]
LGTM, thank you for the contribution @angelosilvestre.

**Pull request merged**
